File: api/src/models/contracts/solutions.py
"""Pydantic types for Solutions — installable surfaces (success-criteria §3)."""

from __future__ import annotations

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field

SolutionScope = Literal["org", "global"]


class SolutionBase(BaseModel):
    slug: str = Field(min_length=1, max_length=255, description="Definition identity (shared across installs)")
    name: str = Field(min_length=1, max_length=255)
    scope: SolutionScope = "org"
    global_repo_access: bool = False
    git_connected: bool = False
    git_repo_url: str | None = None


class SolutionCreate(SolutionBase):
    """Create-shape for an install.

    For ``scope=org`` the install's org is taken from the caller's context (or
    an explicit ``organization_id`` for cross-org admins); ``scope=global``
    means ``organization_id IS NULL``.
    """

    organization_id: UUID | None = None


class SolutionUpdate(BaseModel):
    """Partial-update (PATCH) of an install's INSTALL-LOCAL fields only.

    ``slug`` is identity and is NOT editable here. Portable content
    (workflows/apps/forms/agents/tables/config declarations) is owned by the
    bundle/git and is read-only on this surface.

    PATCH semantics: ``organization_id=None`` is a legitimate value (global
    scope), so it is distinguished from "not provided" via
    ``model_fields_set`` — the endpoint applies only fields present in the
    request (``model_dump(exclude_unset=True)``).
    """

    name: str | None = None
    organization_id: UUID | None = None
    global_repo_access: bool | None = None
    git_connected: bool | None = None
    git_repo_url: str | None = None


class Solution(BaseModel):
    """Read-shape returned by REST.

    ``scope`` is DERIVED from ``organization_id`` (NULL == global), not stored
    on the ORM row — so it always reflects the install's true scope.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name: str
    organization_id: UUID | None = None
    global_repo_access: bool = False
    git_connected: bool = False
    git_repo_url: str | None = None
    # Version bookkeeping (Task 20): the deployed bundle's declared version and
    # what the last version-changing deploy replaced. Deploy-recorded, not
    # caller-settable — version rides in the BUNDLE (descriptor), not a request.
    version: str | None = None
    upgraded_from_version: str | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def scope(self) -> SolutionScope:
        return "org" if self.organization_id is not None else "global"


class SolutionsList(BaseModel):
    solutions: list[Solution] = Field(default_factory=list)


class SolutionConfigStatus(BaseModel):
    """A config DECLARATION on an install, paired with whether a value is set in
    the install's org scope (values are instance-owned Config rows, never part of
    the declaration)."""

    id: UUID
    key: str
    type: str
    required: bool
    description: str | None = None
    value_set: bool


class SolutionEntitySummary(BaseModel):
    """Lightweight (id, name) entry for an owned entity — the detail UI links by id."""

    id: UUID
    name: str


class SolutionEntities(BaseModel):
    """Everything one install owns + its config declaration/value status."""

    solution: Solution
    workflows: list[SolutionEntitySummary] = Field(default_factory=list)
    apps: list[SolutionEntitySummary] = Field(default_factory=list)
    forms: list[SolutionEntitySummary] = Field(default_factory=list)
    agents: list[SolutionEntitySummary] = Field(default_factory=list)
    tables: list[SolutionEntitySummary] = Field(default_factory=list)
    configs: list[SolutionConfigStatus] = Field(default_factory=list)
    required_configs_unset: list[str] = Field(default_factory=list)


class SolutionEntityDiff(BaseModel):
    """Added/removed display names for ONE entity type in an upgrade preview.

    Identity is the deployer's per-install uuid5 remap (``solution_entity_id``),
    so "kept" (in neither list) means the deploy would UPDATE that row in place.
    """

    added: list[str] = Field(default_factory=list)
    removed: list[str] = Field(default_factory=list)


class SolutionConfigSchemaState(BaseModel):
    """The compared portion of a config declaration (type + required)."""

    type: str
    required: bool


class SolutionConfigSchemaChange(BaseModel):
    """One config declaration whose type/required changed between versions."""

    model_config = ConfigDict(populate_by_name=True)

    key: str
    from_: SolutionConfigSchemaState = Field(alias="from")
    to: SolutionConfigSchemaState


class SolutionConfigSchemaDiff(BaseModel):
    """Config DECLARATION diff (by key) for an upgrade preview."""

    added: list[str] = Field(default_factory=list)
    removed: list[str] = Field(default_factory=list)
    changed: list[SolutionConfigSchemaChange] = Field(default_factory=list)


class SolutionUpgradeDiff(BaseModel):
    """What deploying the previewed zip would change on the existing install."""

    workflows: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    tables: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    forms: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    agents: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    apps: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    config_schemas: SolutionConfigSchemaDiff = Field(default_factory=SolutionConfigSchemaDiff)


class SolutionExistingInstall(BaseModel):
    """The install a previewed zip would UPGRADE (matched by slug + scope)."""

    id: UUID
    name: str
    version: str | None = None


class SolutionInstallPreview(BaseModel):
    """Parse-only preview of a Solution install zip — what it would create + its
    declared configs. Nothing is persisted by the preview endpoint.

    When an install already exists for the zip's slug at the requested scope,
    ``existing_install`` + ``diff`` describe the upgrade the install would
    perform (Task 22) — drag-drop routes to UPGRADE, never a second install."""

    slug: str | None = None
    name: str | None = None
    scope: SolutionScope | None = None
    version: str | None = None
    workflows: list[dict[str, Any]] = Field(default_factory=list)
    tables: list[dict[str, Any]] = Field(default_factory=list)
    apps: list[dict[str, Any]] = Field(default_factory=list)
    forms: list[dict[str, Any]] = Field(default_factory=list)
    agents: list[dict[str, Any]] = Field(default_factory=list)
    config_schemas: list[dict[str, Any]] = Field(default_factory=list)
    existing_install: SolutionExistingInstall | None = None
    diff: SolutionUpgradeDiff | None = None


class SolutionDeployRequest(BaseModel):
    """Full-replace deploy bundle for one install.

    ``python_files`` maps relative paths (e.g. ``workflows/w.py``,
    ``modules/x.py``) to UTF-8 source text, installed verbatim under the
    install's ``_solutions/{id}/`` prefix. ``workflows`` are manifest-shaped
    entity dicts to upsert (apps/forms/agents/tables join in later sub-plans).
    Deploy is non-interactive by contract — it always applies the full bundle.
    """

    python_files: dict[str, str] = Field(default_factory=dict)
    workflows: list[dict[str, Any]] = Field(default_factory=list)
    tables: list[dict[str, Any]] = Field(default_factory=list)
    # Each app: {id, slug, name, app_model, dependencies, access_level,
    # src_files: {rel: text} | dist_files: {rel: text}}. dist_files is the
    # disconnected fast-path (skip the server build).
    apps: list[dict[str, Any]] = Field(default_factory=list)
    # Each form: {id, name, description?, workflow_id?, fields: [...]}.
    forms: list[dict[str, Any]] = Field(default_factory=list)
    # Each agent: {id, name, system_prompt, description?, channels?, llm_model?}.
    agents: list[dict[str, Any]] = Field(default_factory=list)
    # Each config schema: {id, key, type, required, description?, default?, position}.
    # DECLARATIONS only — never a value (values are instance-owned Config rows).
    config_schemas: list[dict[str, Any]] = Field(default_factory=list)
    # The bundle's declared version (bifrost.solution.yaml ``version:``).
    # Recorded on the install; an older version than installed is refused
    # unless ``force`` is set (Task 20 downgrade gate).
    version: str | None = None
    force: bool = False


class SolutionDeleteSummary(BaseModel):
    """Counts of what a DELETE did. Pure-code entities (workflows/apps/forms/
    agents) and the install's config DECLARATIONS are deleted via DB cascade.
    Data-bearing entities are ORPHANED, not deleted: owned tables (and their
    documents) are detached and survive as ordinary org tables, and the
    install's config VALUES are stamped with orphan provenance and survive.
    The UI echoes these back to the operator."""

    solution_id: UUID
    workflows_deleted: int = 0
    apps_deleted: int = 0
    forms_deleted: int = 0
    agents_deleted: int = 0
    config_declarations_deleted: int = 0
    tables_orphaned: int = 0
    config_values_orphaned: int = 0


class SolutionDeployResponse(BaseModel):
    solution_id: UUID
    workflows_upserted: int = 0
    workflows_deleted: int = 0
    tables_upserted: int = 0
    tables_deleted: int = 0
    apps_upserted: int = 0
    apps_deleted: int = 0
    forms_upserted: int = 0
    forms_deleted: int = 0
    agents_upserted: int = 0
    agents_deleted: int = 0

File: client/src/services/solutions.ts
import { apiClient, authFetch } from "@/lib/api-client";
import { getErrorMessage } from "@/lib/api-error";
import type { components } from "@/lib/v1";

export type Solution = components["schemas"]["Solution"];
export type SolutionsList = components["schemas"]["SolutionsList"];
export type SolutionEntities = components["schemas"]["SolutionEntities"];
export type SolutionInstallPreview =
	components["schemas"]["SolutionInstallPreview"];
export type SolutionExistingInstall =
	components["schemas"]["SolutionExistingInstall"];
export type SolutionUpgradeDiff =
	components["schemas"]["SolutionUpgradeDiff"];
export type SolutionDeleteSummary =
	components["schemas"]["SolutionDeleteSummary"];
export type SolutionUpdate = components["schemas"]["SolutionUpdate"];

interface RequestOptions {
	signal?: AbortSignal;
}

export async function listSolutions(
	options: RequestOptions = {},
): Promise<SolutionsList> {
	const { signal } = options;
	const { data, error } = await apiClient.GET("/api/solutions", { signal });
	if (error) throw new Error(getErrorMessage(error, "Failed to list solutions"));
	return data;
}

export async function getSolution(
	solutionId: string,
	options: RequestOptions = {},
): Promise<Solution> {
	const { signal } = options;
	const { data, error } = await apiClient.GET("/api/solutions/{solution_id}", {
		params: { path: { solution_id: solutionId } },
		signal,
	});
	if (error) throw new Error(getErrorMessage(error, "Failed to get solution"));
	return data;
}

export async function getSolutionEntities(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionEntities> {
	const { signal } = options;
	const { data, error } = await apiClient.GET(
		"/api/solutions/{solution_id}/entities",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to get solution entities"));
	}
	return data;
}

export async function updateSolution(
	solutionId: string,
	update: SolutionUpdate,
	options: RequestOptions = {},
): Promise<Solution> {
	const { signal } = options;
	const { data, error } = await apiClient.PATCH(
		"/api/solutions/{solution_id}",
		{ params: { path: { solution_id: solutionId } }, body: update, signal },
	);
	if (error) throw new Error(getErrorMessage(error, "Failed to update solution"));
	return data;
}

/**
 * Set a config VALUE for a Solution install's org scope. Config values are
 * instance-owned `Config` rows (never part of the portable declaration), so we
 * write them through the existing `/api/config` endpoint scoped to the install's
 * organization. `organizationId` is the install's org (`null` for a global install).
 */
export async function setSolutionConfig(
	params: {
		key: string;
		value: string;
		type: components["schemas"]["ConfigType"];
		organizationId: string | null;
	},
	options: RequestOptions = {},
): Promise<void> {
	const { key, value, type, organizationId } = params;
	const { error } = await apiClient.POST("/api/config", {
		body: { key, value, type, organization_id: organizationId },
		signal: options.signal,
	});
	if (error) throw new Error(getErrorMessage(error, "Failed to save config value"));
}

export async function deleteSolution(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionDeleteSummary> {
	const { signal } = options;
	const { data, error } = await apiClient.DELETE(
		"/api/solutions/{solution_id}",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) throw new Error(getErrorMessage(error, "Failed to delete solution"));
	return data;
}

/**
 * Download the install's workspace zip (the exact bundle its last write
 * produced). Returns the blob + the server-chosen filename so the caller can
 * trigger a browser download.
 */
export async function exportSolution(
	solutionId: string,
): Promise<{ blob: Blob; filename: string }> {
	const response = await authFetch(`/api/solutions/${solutionId}/export`);
	if (!response.ok) {
		throw new Error(
			await parseUploadError(response, "Failed to export solution"),
		);
	}
	const disposition = response.headers.get("Content-Disposition") ?? "";
	const match = /filename="([^"]+)"/.exec(disposition);
	return {
		blob: await response.blob(),
		filename: match?.[1] ?? `solution-${solutionId}.zip`,
	};
}

async function parseUploadError(
	response: Response,
	fallback: string,
): Promise<string> {
	const body = await response.json().catch(() => ({}));
	if (body && typeof body.detail === "string") {
		return body.detail;
	}
	return fallback;
}

/**
 * Preview a Solution install zip (parse-only). Posts a multipart `file` and an
 * optional `organization_id` (empty/absent = global) so the server can match
 * an existing install at that scope and return `existing_install` + `diff`.
 */
export async function previewInstall(
	file: File,
	params: { organizationId?: string } = {},
	options: RequestOptions = {},
): Promise<SolutionInstallPreview> {
	const formData = new FormData();
	formData.append("file", file);
	formData.append("organization_id", params.organizationId ?? "");

	const response = await authFetch("/api/solutions/install/preview", {
		method: "POST",
		body: formData,
		signal: options.signal,
	});
	if (!response.ok) {
		throw new Error(
			await parseUploadError(
				response,
				`Failed to preview install: ${response.statusText}`,
			),
		);
	}
	return response.json();
}

/**
 * Install a Solution zip. Posts a multipart `file`, optional `organization_id`
 * (empty string installs globally), and `config_values` (JSON-encoded map).
 * Pass `force: true` to override the server's downgrade guard (409 when the
 * package version is older than the installed version).
 */
export async function installSolution(
	params: {
		file: File;
		organizationId?: string;
		configValues?: Record<string, unknown>;
		force?: boolean;
	},
	options: RequestOptions = {},
): Promise<Solution> {
	const { file, organizationId, configValues, force } = params;
	const formData = new FormData();
	formData.append("file", file);
	formData.append("organization_id", organizationId ?? "");
	formData.append("config_values", JSON.stringify(configValues ?? {}));

	const url = force
		? "/api/solutions/install?force=true"
		: "/api/solutions/install";
	const response = await authFetch(url, {
		method: "POST",
		body: formData,
		signal: options.signal,
	});
	if (!response.ok) {
		throw new Error(
			await parseUploadError(
				response,
				`Failed to install solution: ${response.statusText}`,
			),
		);
	}
	return response.json();
}
